README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# mlsjunkgen <img src="man/figures/logo.png" align="right" height="139" />

<!-- badges: start -->

[![R-CMD-check](https://github.com/scumdogsteev/mlsjunkgen/workflows/R-CMD-check/badge.svg)](https://github.com/scumdogsteev/mlsjunkgen/actions)
[![CRAN\_Status\_Badge](https://www.r-pkg.org/badges/version/mlsjunkgen)](https://cran.r-project.org/package=mlsjunkgen)
[![Coverage
Status](https://img.shields.io/coveralls/scumdogsteev/mlsjunkgen.svg)](https://coveralls.io/github/scumdogsteev/mlsjunkgen)
[![RStudio CRAN mirror
downloads](https://cranlogs.r-pkg.org/badges/grand-total/mlsjunkgen)](https://cran.r-project.org/package=mlsjunkgen)
<!-- badges: end -->

### Background

**`mlsjunkgen`** is a pseudo-random number generator.

#### Algorithm

For any seed values of w, x, y, z:

r<sub>i</sub> = 5.980217w<sup>2</sup> + 9.446377x<sup>0.25</sup> +
4.81379y<sup>0.33</sup> + 8.91197z<sup>0.5</sup>

r<sub>i</sub> = r<sub>i</sub> - Int(r<sub>i</sub>)

For r<sub>i</sub>+1:

w = x

x = y

y = z

z = r<sub>i</sub>

#### Analysis

This generator tends to do well with various tests for randomness (K-S,
Chi Square, test for runs up and down). It may not perform as well on
other tests (e.g., tests for runs above and below the mean), but that
could relate to my choice of seeds. As a point of reference, the period
of Excel’s built-in random number generator is 16,777,216 and the MLS
Junk Generator’s period is something greater than 9.9 billion (the point
at which I gave up on trying to determine it).

### Installation

-   `mlsjunkgen` is available [on
    CRAN](https://cran.r-project.org/package=mlsjunkgen) and can be
    installed accordingly:

``` r
install.packages("mlsjunkgen")
library(mlsjunkgen)
```

-   You can also install `mlsjunkgen` from GitHub using the `devtools`
    package:

``` r
install.packages("devtools")
library("devtools")
install_github("scumdogsteev/mlsjunkgen")
library(mlsjunkgen)
```

### Usage

The package consists of four functions:

1.  `junkgen` - generates a pseudo-random number from user-specified
    seeds
2.  `mlsjunkgenv` - generates a vector of pseudo-random numbers by
    calling `junkgen` a user-specified number of times
3.  `mlsjunkgend` - generates a data frame of pseudo-random numbers by
    calling `junkgen` a user-specified number of times
4.  `mlsjunkgenm` - generates a user-specified size matrix of
    pseudo-random numbers by calling `mlsjunkgenv` and assigning the
    results to a matrix

#### Examples

**`junkgen`** generates a single pseudo-random number based on four
user-specified seeds:

``` r
w <- 1
x <- 2
y <- 3
z <- 4
junkgen(w = w, x = x, y = y, z = z)
#> [1] 0.9551644
```

**`mlsjunkgenv`** generates a vector containing a stream of `n` (default
= 1) user-specified pseudo-random numbers based on four user-specified
seeds rounded to a specified (default = 5) number of decimal places:

``` r
mlsjunkgenv(n = 10, w = w, x = x, y = y, z = z, round = 2)
#>  [1] 0.96 0.67 0.21 0.34 0.12 0.56 0.59 0.11 0.34 0.70
```

The same example with default rounding:

``` r
mlsjunkgenv(n = 10, w = w, x = x, y = y, z = z)
#>  [1] 0.95516 0.66908 0.21235 0.34488 0.11995 0.56398 0.59235 0.11432 0.33525
#> [10] 0.70271
```

**`mlsjunkgend`** generates a data frame containing a stream of `n`
user-specified pseudo-random numbers based on four user-specified seeds:

``` r
mlsjunkgend(n = 10, w = w, x = x, y = y, z = z, round = 2)
#>      RN
#> 1  0.96
#> 2  0.67
#> 3  0.21
#> 4  0.34
#> 5  0.12
#> 6  0.56
#> 7  0.59
#> 8  0.11
#> 9  0.34
#> 10 0.70
```

The same example with default rounding:

``` r
mlsjunkgend(n = 10, w = w, x = x, y = y, z = z)
#>         RN
#> 1  0.95516
#> 2  0.66908
#> 3  0.21235
#> 4  0.34488
#> 5  0.11995
#> 6  0.56398
#> 7  0.59235
#> 8  0.11432
#> 9  0.33525
#> 10 0.70271
```

**`mlsjunkgenm`** generates a matrix of user-specified size containing a
stream of pseudo-random numbers based on four user-specified seeds:

``` r
mlsjunkgenm(nrow = 5, ncol = 5, w = w, x = x, y = y, z = z, round = 3)
#>       [,1]  [,2]  [,3]  [,4]  [,5]
#> [1,] 0.955 0.564 0.418 0.052 0.020
#> [2,] 0.669 0.592 0.313 0.663 0.110
#> [3,] 0.212 0.114 0.920 0.802 0.685
#> [4,] 0.345 0.335 0.379 0.160 0.286
#> [5,] 0.120 0.703 0.280 0.586 0.452
```


NEWS.md
# mlsjunkgen 0.1.2 - Double T

- add rmarkdown as a dependency for vignette
- update project and other URLs

# mlsjunkgen 0.1.1 - Masked Rider

- fix typos in README and vignette
- correct README and vignette examples

# mlsjunkgen 0.1.0 - Raider Red

- first version of mlsjunkgen package


tests/testthat.R
library(testthat)
library(mlsjunkgen)

test_check("mlsjunkgen")


tests/testthat/test-junkgen.R
library(mlsjunkgen)

test_that("junkgen returns the correct psuedo-random number", {
    expect_equal(round(junkgen(w = 1, x = 2, y = 3, z = 4), 5), 0.95516)
    expect_equal(round(junkgen(w = 2, x = 3, y = 4, z = 5), 5), 0.88696)
    expect_equal(round(junkgen(w = 0.886956, x = 0.886956, y = 0.9551644,
                               z = 0.9551644), 5), 0.32323)
    expect_error(junkgen(w = "X", x = 2, y = 3, z = 4), "Invalid input.  Please ensure all seeds are numeric.")
})


tests/testthat/test-mlsjunkgend.R
library(mlsjunkgen)

dat <- data.frame(0.95516)
names(dat) <- "RN"
dat2 <- data.frame(c(0.95516, 0.66908, 0.21235, 0.34488, 0.11995))
names(dat2) <- "RN"
dat3 <- data.frame(c(0.887, 0.762, 0.214, 0.197, 0.375, 0.168))
names(dat3) <- "RN"


test_that("mlsjunkgend returns the correct data frame", {
    expect_equal(mlsjunkgend(w = 1, x = 2, y = 3, z = 4), dat)
    expect_equal(mlsjunkgend(n = 5, w = 1, x = 2, y = 3, z = 4), dat2)
    expect_equal(mlsjunkgend(n = 6, w = 2, x = 3, y = 4, z = 5, round = 3),
                 dat3)
    expect_error(mlsjunkgend(n = "X", w = 1, x = 2, y = 3, z = 4), "Invalid input.  Please ensure n is numeric.")
    expect_error(mlsjunkgend(n = 5, w = "X", x = 2, y = 3, z = 4), "Invalid input.  Please ensure all seeds are numeric.")
})


R/junkgen.R
#'
#' Generate a single pseudo-random number using the MLS Junk Generator algorithm
#'
#' Based on user input seeds, this function generates a pseudo-random number.
#' This is called by the mlsjunkgen package's other functions to generate a
#' pseudo-random number stream.
#'
#' @param w the first seed required by the MLS Junk Generator algorithm
#' @param x the first seed required by the MLS Junk Generator algorithm
#' @param y the first seed required by the MLS Junk Generator algorithm
#' @param z the first seed required by the MLS Junk Generator algorithm
#'
#' @export
#'
#' @examples
#' # Generate a pseudo-random number with user-specified seeds
#'
#' w <- 1
#' x <- 2
#' y <- 3
#' z <- 4
#' junkgen(w = w, x = x, y = y, z = z) # returns "[1] 0.9551644"
#'
#' @return A numeric vector containing a single pseudo-random number

junkgen <- function (w, x, y, z) {
    if (is.numeric(w) & is.numeric(x) & is.numeric(y) & is.numeric(z)) {
        r <- 5.980217 * (w ^ 2) + 9.446377 * (x ^ 0.25) + 4.81379 * (y ^ 0.33) +
         8.91197 * (z ^ 0.5)
        ri <- r - trunc(r)
    } else { stop("Invalid input.  Please ensure all seeds are numeric.") }
    return(ri)
}


R/mlsjunkgend.R
#'
#' Generate a data frame of pseudo-random numbers using the MLS Junk Generator algorithm
#'
#' Based on user input seeds, this function generates a data frame of n
#' pseudo-random numbers and names the column containing these as "RN" for "random
#' numbers."  This is achieved by calling junkgen.
#'
#' @param n the number of pseudo-random numbers to generate; defaults to 1
#' @param w the first seed required by the MLS Junk Generator algorithm
#' @param x the first seed required by the MLS Junk Generator algorithm
#' @param y the first seed required by the MLS Junk Generator algorithm
#' @param z the first seed required by the MLS Junk Generator algorithm
#' @param round the number of decimal places to which to round the pseudo-random numbers; default = 5
#'
#' @export
#'
#' @examples
#' # Generate a pseudo-random number data frame with 10 observations from user-specified seeds
#'
#' w <- 1
#' x <- 2
#' y <- 3
#' z <- 4
#'
#' mlsjunkgend(n = 10, w = w, x = x, y = y, z = z) # returns a data frame of 10 observations
#'
#' # Specifying different values for n and round
#'
#' mlsjunkgend(n = 5, w = w, x = x, y = y, z = z, round = 2)
#' # returns a data frame identical to the above example but with only 5 observations
#' # rounded to 2 decimal places
#'
#' # using the default value of n (1) is identical to assigning the rounded result of
#' # junkgen to a data frame of 1 observation
#'
#' round(junkgen(w = w, x = x, y = y, z = z), 5) # returns "[1] 0.95516"
#' mlsjunkgend(w = w, x = x, y = y, z = z)
#' # returns the following:
#' #        RN
#' # 1 0.95516
#'
#' @return A numeric vector containing a single pseudo-random number

mlsjunkgend <- function(n = 1, w, x, y, z, round = 5) {
    if (is.numeric(n)) { mls <- data.frame()
                         for (i in 1:n) { ri <- junkgen(w, x, y, z)
                                          mls <- rbind(mls, round(ri, round))
                                          w <- x
                                          x <- y
                                          y <- z
                                          z <- ri
                                        }
                         names(mls) <- "RN"
                         return(mls)
                       }
    else { stop("Invalid input.  Please ensure n is numeric.") }
}


R/mlsjunkgenm.R
#'
#' Generate a matrix of pseudo-random numbers using the MLS Junk Generator algorithm
#'
#' Based on user input seeds, this function generates a vector of n
#' pseudo-random numbers by calling mlsjunkgenv which in turn calls junkgen.
#'
#' @param nrow the number of rows for the matrix; defaults to 1
#' @param ncol the number of columns for the matrix; defaults to 1
#' @param w the first seed required by the MLS Junk Generator algorithm
#' @param x the first seed required by the MLS Junk Generator algorithm
#' @param y the first seed required by the MLS Junk Generator algorithm
#' @param z the first seed required by the MLS Junk Generator algorithm
#' @param round the number of decimal places to which to round the pseudo-random numbers; default = 5
#'
#' @export
#'
#' @examples
#' # Generate a 4x4 matrix of pseudo-random numbers with user-specified seeds
#'
#' w <- 1
#' x <- 2
#' y <- 3
#' z <- 4
#'
#' mlsjunkgenm(nrow = 4, ncol = 4, w = w, x = x, y = y, z = z) # returns a 4x4 matrix
#'
#' # the sixteen values in the above matrix are equivalent to the following call
#' # to mlsjunkgenv
#'
#' mlsjunkgenv(n = 16, w = w, x = x, y = y, z = z)
#'
#' # matrices need not be square
#' # this returns a 3x2 matrix of pseudo-random numbers with 2 decimal places
#' mlsjunkgenm(nrow = 3, ncol = 2, w = w, x = x, y = y, z = z, round = 2)
#'
#' # using the default value of n (1) generates a 1x1 matrix the value of which
#' # is identical to running junkgen and rounding the result to 5 decimal places
#'
#' round(junkgen(w = w, x = x, y = y, z = z), 5) # returns "[1] 0.95516"
#' mlsjunkgenv(w = w, x = x, y = y, z = z) # returns a 1x1 matrix with single element = "0.95516"
#'
#' @return A numeric vector containing a single pseudo-random number

mlsjunkgenm <- function(nrow = 1, ncol = 1, w, x, y, z, round = 5) {
    if (is.numeric(nrow) & is.numeric(ncol)) {
        matrix(mlsjunkgenv(nrow * ncol, w, x, y, z, round), nrow = nrow,
               ncol = ncol)
    } else { stop("Invalid input.  Please ensure nrow and ncol are numeric.") }
}


R/mlsjunkgenv.R
#'
#' Generate a vector of pseudo-random numbers using the MLS Junk Generator algorithm
#'
#' Based on user input seeds, this function generates a vector of n
#' pseudo-random numbers by calling junkgen.
#'
#' @param n the number of pseudo-random numbers to generate; defaults to 1
#' @param w the first seed required by the MLS Junk Generator algorithm
#' @param x the first seed required by the MLS Junk Generator algorithm
#' @param y the first seed required by the MLS Junk Generator algorithm
#' @param z the first seed required by the MLS Junk Generator algorithm
#' @param round the number of decimal places to which to round the pseudo-random numbers; default = 5
#'
#' @export
#'
#' @examples
#' # Generate a pseudo-random number stream of length 5 with user-specified seeds
#'
#' w <- 1
#' x <- 2
#' y <- 3
#' z <- 4
#'
#' # the following call returns "[1] 0.95516 0.66908 0.21235 0.34488 0.11995"
#' mlsjunkgenv(n = 5, w = w, x = x, y = y, z = z)
#'
#' # Specifying different values for n and round
#'
#' mlsjunkgenv(n = 3, w = w, x = x, y = y, z = z, round = 2) # returns "[1] 0.96 0.67 0.21"
#'
#' # using the default value of n (1) is identical to running junkgen and rounding
#' # the result to 5 decimal places
#'
#' round(junkgen(w = w, x = x, y = y, z = z),5) # returns "[1] 0.95516"
#' mlsjunkgenv(w = w, x = x, y = y, z = z) # returns "[1] 0.95516"
#'
#' @return A numeric vector containing a single pseudo-random number

mlsjunkgenv <- function(n = 1, w, x, y, z, round = 5) {
    if (is.numeric(n)) { mls <- numeric()
                         for (i in 1:n) { ri <- junkgen(w, x, y, z)
                                          mls <- c(mls, round(ri, round))
                                          w <- x
                                          x <- y
                                          y <- z
                                          z <- ri
                         }
                         return(mls)
                    }
    else { stop("Invalid input.  Please ensure n is numeric.") }
}
